### PR TITLE
[PubGrub] Avoid resolving package versions twice in presence of prebuilt libraries

### PR DESCRIPTION
**Pull request by xedin**
Integrate information about versions of prebuilt libraries into the solver
and prioritize these versions when making decisions if they meet all of the
requirements of undecided terms.

This helps solver to avoid fetching and at the same time allows partual
matches on prebuilt libraries with fallback to package provided versions.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by MaxDesiatov** [comment]
@swift-ci test macos

**Review by MaxDesiatov — APPROVED** [review]
LGTM, only an indentation nit

File: Tests/PackageGraphTests/PubgrubTests.swift [hunk 1]
```diff
@@ -3299,13 +3392,22 @@ class DependencyGraphBuilder {
     }
 
 
-    func create(pins: PinsStore.Pins = [:], delegate: DependencyResolverDelegate? = .none) -> PubGrubDependencyResolver {
+    func create(pins: PinsStore.Pins = [:],
+                availableLibraries: [LibraryMetadata] = [],
+                delegate: DependencyResolverDelegate? = .none
+    ) -> PubGrubDependencyResolver {
```
[the remark this answers is not in the document]
Comment (xedin) [reply Tests/PackageGraphTests/PubgrubTests.swift:3398]: I'm going to run `swiftformat` on all of the files I changed and check that in as a separate commit.

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Comment by xedin** [comment]
@swift-ci test macos

**Comment by xedin** [comment]
@swift-ci test macos

**Pull request merged**